source/MaterialXContrib/MaterialXNode/MaterialXSurfaceOverride.cpp

```diff
@@ -0,0 +1,282 @@
+#include "MaterialXSurfaceOverride.h"
+#include "MaterialXNode.h"
+#include "Plugin.h"
+#include "Util.h"
+
+#include <MaterialXGenGlsl/GlslShaderGenerator.h>
+#include <MaterialXGenShader/HwShaderGenerator.h>
+#include <MaterialXGenShader/GenContext.h>
+#include <MaterialXGenShader/Util.h>
+#include <MaterialXFormat/XmlIo.h>
+#include <MaterialXRender/StbImageLoader.h>
+
+#include <maya/MFnDependencyNode.h>
+#include <maya/MFragmentManager.h>
+#include <maya/MRenderUtil.h>
+#include <maya/MShaderManager.h>
+#include <maya/MTextureManager.h>
+
+#include <fstream>
+
+const MString
+    MaterialXSurfaceOverride::REGISTRANT_ID = "materialXSurface",
+    MaterialXSurfaceOverride::DRAW_CLASSIFICATION = "drawdb/shader/surface/materialX";
+
+MHWRender::MPxShadingNodeOverride* MaterialXSurfaceOverride::creator(const MObject& obj)
+{
+	std::cout.rdbuf(std::cerr.rdbuf());
+	return new MaterialXSurfaceOverride(obj);
+}
+
+MaterialXSurfaceOverride::MaterialXSurfaceOverride(const MObject& obj)
+	: MPxShadingNodeOverride(obj)
+	, _object(obj)
+{
+	MStatus status;
+	MFnDependencyNode depNode(_object, &status);
+	if (status)
+	{
+		depNode.findPlug(MaterialXNode::DOCUMENT_ATTRIBUTE_LONG_NAME, false, &status).getValue(_documentContent);
+		depNode.findPlug(MaterialXNode::ELEMENT_ATTRIBUTE_LONG_NAME, false, &status).getValue(_element);
+	}
+
+    try
+    {
+        // Create document
+        _document = MaterialX::createDocument();
+        MaterialX::readFromXmlString(_document, _documentContent.asChar());
+
+        // Load libraries - why is this done twice. Once in the command and once here??
+        MaterialX::FilePath libSearchPath = Plugin::instance().getLibrarySearchPath();
+        const MaterialX::StringVec libraries = { "stdlib", "pbrlib", "bxdf", "stdlib/genglsl", "pbrlib/genglsl" };
+        MaterialX::loadLibraries(libraries, libSearchPath, _document);
+
+        std::cout.rdbuf(std::cerr.rdbuf());
+
+        std::vector<MaterialX::GenContext*> contexts;
+        MaterialX::GenContext* glslContext = new MaterialX::GenContext(MaterialX::GlslShaderGenerator::create());
+
+        // Stop emission of environment map lookups.
+        glslContext->getOptions().hwSpecularEnvironmentMethod = MaterialX::SPECULAR_ENVIRONMENT_NONE;
+        glslContext->registerSourceCodeSearchPath(libSearchPath);
+        contexts.push_back(glslContext);

```

Comment:
I think this glslContext pointer is leaked?

Reply:
Right. I fixed this for MaterialXTextureOverride but left MaterialXSurfaceOverride as is because @bernardkwok was experimenting with it. I guess a lot of the code between the two override should eventually move to MaterialXData to avoid diplication.